File: akka_essentials/src/main/java/chapter2java8/actor/MapActor.java
package chapter2java8.actor;

import akka.actor.AbstractActor;
import akka.japi.pf.ReceiveBuilder;
import chapter2.message.MapData;
import chapter2.message.WordCount;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.StringTokenizer;

public class MapActor extends AbstractActor {

    private static final String[] STOP_WORDS = {"a", "am", "an", "and", "are", "as", "at",
            "be", "do", "go", "if", "in", "is", "it", "of", "on", "the", "to"};
    private static final List<String> STOP_WORDS_LIST = Arrays.asList(STOP_WORDS);

    public MapActor() {
        receive(ReceiveBuilder.
                match(String.class, str -> sender().tell(evaluateExpression(str), self()))
                .build());
    }

    private MapData evaluateExpression(String line) {
        List<WordCount> dataList = new ArrayList<>();
        StringTokenizer parser = new StringTokenizer(line);
        while (parser.hasMoreTokens()) {
            String word = parser.nextToken().toLowerCase();
            if (!STOP_WORDS_LIST.contains(word)) {
                dataList.add(new WordCount(word, 1));
            }
        }
        return new MapData(dataList);
    }
}

File: akka_essentials/src/main/scala/essentials/actor/scala/chapter2/actor/MapActor.scala
package essentials.actor.scala.chapter2.actor

import akka.actor.Actor
import essentials.actor.scala.chapter2.{MapData, WordCount}

import scala.collection.mutable.ArrayBuffer

class MapActor extends Actor {

  val STOP_WORDS_LIST = List("a", "am", "an", "and", "are", "as", "at", "be", "do", "go", "if", "in", "is", "it", "of", "on", "the", "to")

  def receive: Receive = {
    case message: String => sender ! evaluateExpression(message)
  }

  def evaluateExpression(line: String): MapData = MapData {
    line.split("""\s+""").foldLeft(ArrayBuffer.empty[WordCount]) {
      (index, word) =>
        if (!STOP_WORDS_LIST.contains(word.toLowerCase))
          index += WordCount(word.toLowerCase, 1)
        else
          index
    }
  }

}
